pynest/nest/lib/hl_api_exceptions.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+#
+# hl_api_exceptions.py
+#
+# This file is part of NEST.
+#
+# Copyright (C) 2004 The NEST Initiative
+#
+# NEST is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# NEST is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+
+class NESTMappedException(type):
+    """Metaclass for exception namespace that dynamically creates exception classes.
+
+    If a class (self) of this (meta)-type has an unknown attribute requested, __getattr__ defined
+    below gets called, creating a class with that name (the error name) and with an __init__ taking
+    commandname and errormessage (as created in the source) which is a closure on the parent and
+    errorname as well, with a parent of default type (self.default_parent) or
+    self.parents[errorname] if defined. """
+
+    def __getattr__(self, errorname):
+        """Creates a class of type "errorname" which is a child of self.default_parent or
+        self.parents[errorname] if one is defined.
+
+        This __getattr__ function also stores the class permanently as an attribute of self for
+        re-use where self is actually the class that triggered the getattr (the class that
+        NESTMappedException is a metaclass of). """
+
+        # Dynamic class construction, first check if we know its parent
+        if errorname in self.parents:
+            parent = getattr(self, self.parents[errorname])
+        else:  # otherwise, get the default (SLIException)
+            parent = self.default_parent
+
+        # and now dynamically construct the new class
+        # not NESTMappedException, since that would mean the metaclass would let the new class inherit
+        # this __getattr__, allowing unintended dynamic construction of attributes
+        newclass = type(
+            self.__name__ + '.' + errorname,
+            (parent,),
+            {
+                '__init__': self.init(parent, errorname),
+                '__doc__':
+                """Dynamically created exception {} from {}.
+
+                Created for the namespace: {}.
+                Parent exception: {}.
+                """.format(errorname, self.source, self.__name__, parent.__name__)
+            }
+        )
+
+        # Cache for reuse: __getattr__ should now not get called if requested again
+        setattr(self, errorname, newclass)
+
+        # And now we return the exception
+        return newclass
+
+
+class NESTErrors(metaclass=NESTMappedException):
```

Comment:
switch to the standard python 2/3 compatible version:
```suggestion
class NESTErrors:
    __metaclass__ = NESTMappedException
```